### ISSUE TITLE
Pixels shifts (EXIF orientation images)

### ISSUE THREAD

**Issue by kleisauke** [libvips/libvips#703]
Discovered a strange bug when testing the auto rotation of EXIF orientation images, it looks like the pixels are shifting for some images.

Test images:
https://github.com/recurser/exif-orientation-examples

Test case:
[bug.zip](https://github.com/jcupitt/libvips/files/1194227/bug.zip)

Usage:
- `composer install`
- `php test.php`
- see the output folder

Output as GIF slideshow:
![test](https://user-images.githubusercontent.com/12746591/28879185-caad1da4-77a1-11e7-9200-077ec1e0f610.gif)
<sup>It's GIF so horrible quality, but you clearly see the (random?) pixel shifts by the portrait images.</sup>

Tested with libvips `8.6.0-alpha1` and `8.5.7`, the output is the same.

**Comment by jcupitt** [comment]
Wow that is strange. I'll have a look.

**Comment by jcupitt** [comment]
Back on this after a break. It looks like it's a problem in reduce, I'll try to make a test case.

**Comment by kleisauke** [comment]
I could make a test case with `vips_resize` or `vips_reduce`, if you want.

**Comment by jcupitt** [comment]
I think we'll have to bump this to 8.7, 8.6 is already horribly delayed. What do you think?

**Comment by kleisauke** [comment]
I agree, let's bump this issue to 8.7. It's just a minor issue that I've encountered.

**Comment by kleisauke** [comment]
FWIW, there are fewer pixels shifts with commit https://github.com/jcupitt/libvips/commit/7db13412e553e5f9e59e24f0fe3cfc074be51032 (which uses corner convention for down-sampling).

![pixel-shifts](https://user-images.githubusercontent.com/12746591/33241466-08547510-d2c6-11e7-8776-9e38163ce658.gif)
<sup>Again, it's a GIF image so horrible quality...</sup>

https://github.com/jcupitt/libvips/issues/705 might be relevant here.

**Comment by kleisauke** [comment]
I've set-up a separate repo that reproduces this: https://github.com/kleisauke/vips-issue-703.

Output:
![Portrait](https://github.com/kleisauke/vips-issue-703/raw/master/output/Portrait.webp)
![Landscape](https://github.com/kleisauke/vips-issue-703/raw/master/output/Landscape.webp)
<sup>It's an animated WebP image, the quality should be better now.</sup>

@jcupitt You may bump this to another milestone, debugging this issue is difficult. :sweat_smile:

**Comment by kleisauke** [comment]
Quick update: Corner convention seems to produce more pixel shifts. The output can be viewed here:
https://github.com/kleisauke/vips-issue-703#corner-vs-center-convention-portrait

I also added some debugging notes to quickly reproduce this with the command line:
https://github.com/kleisauke/vips-issue-703#debugging-notes

**Comment by kleisauke** [comment]
This [patch](https://github.com/kleisauke/vips-issue-703/blob/9a9188e1181ca9238e48760ab55cf56fe97ceb14/0001-partially-fix.patch) seems to partially fix this. Output:
![Portrait](https://github.com/kleisauke/vips-issue-703/raw/master/output/Portrait-patch.webp)
![Landscape](https://github.com/kleisauke/vips-issue-703/raw/master/output/Landscape-patch.webp)

The portrait image seems fixed, but the landscape image still produces some pixel shifts.

**Comment by kleisauke** [comment]
I tried to disable the JPEG shrink-on-load feature (commit https://github.com/kleisauke/vips-issue-703/commit/f91737ddc85f90852a6239958704fe753df7b867). It seems to produce more pixel shifts when it's disabled, which makes sense because libvips now have to resize the whole image instead of the last 200%.

Shrink-on-load enabled:
![Landscape](https://github.com/kleisauke/vips-issue-703/raw/c1b3061d546c35fc14531456fef6102273d46da6/output/Landscape.webp)

Shrink-on-load disabled:
![Landscape](https://github.com/kleisauke/vips-issue-703/raw/f91737ddc85f90852a6239958704fe753df7b867/output/Landscape.webp)

With this [patch](https://github.com/kleisauke/vips-issue-703/blob/c1b3061d546c35fc14531456fef6102273d46da6/0001-Partially-fix-for-issue-703.patch) applied it generates no visual difference whether shrink-on-load is enabled or disabled. Unfortunately, it still produces some small pixel shifts.

Shrink-on-load enabled with patch:
![Landscape](https://github.com/kleisauke/vips-issue-703/raw/c1b3061d546c35fc14531456fef6102273d46da6/output-patch/Landscape.webp)

Shrink-on-load disabled with patch:
![Landscape](https://github.com/kleisauke/vips-issue-703/raw/f91737ddc85f90852a6239958704fe753df7b867/output-patch/Landscape.webp)

**Comment by kleisauke** [comment]
I've made another attempt to fix this issue, and I may have found a proper solution for at least `reduceh`.

When I looked at ImageMagick's source code to see how they handle the centre convention, I found something interesting:
https://github.com/ImageMagick/ImageMagick6/blob/8ed8ee76dd5fcfc3d182cf00c5696508425176b4/magick/resize.c#L1752-L1755

From the above lines, I assume that the centre convention (`+ 0.5`) is applied before it's multiplied by the amount of scale. libvips does this after the multiplication, see:
https://github.com/libvips/libvips/blob/10c4831a709063189d67b7d6076b1b26bf62b2b2/libvips/resample/reduceh.cpp#L337-L339

So, I tried to do the same thing:
```diff
@@ -334,9 +324,7 @@ vips_reduceh_gen( VipsRegion *out_region, void *seq,
 
 		q = VIPS_REGION_ADDR( out_region, r->left, r->top + y );
 
-		X = r->left * reduceh->hshrink;
-		if( reduceh->centre )
-			X += 0.5;
+		X = (r->left + offset) * reduceh->hshrink - offset;
 
 		/* We want p0 to be the start (ie. x == 0) of the input 
 		 * scanline we are reading from. We can then calculate the p we
```
(where `offset = 0.5` in case of centre convention)

This seems to solve the problems within `reduceh`. When I tried to do the same within `reducev`:
```diff
@@ -544,8 +545,8 @@ vips_reducev_gen( VipsRegion *out_region, void *vseq,
 	for( int y = 0; y < r->height; y ++ ) { 
 		VipsPel *q = 
 			VIPS_REGION_ADDR( out_region, r->left, r->top + y );
-		const double Y = (r->top + y) * reducev->vshrink + 
-			(reducev->centre ? 0.5 : 0.0); 
+		const double Y = (r->top + y + offset) * reducev->vshrink -
+			offset;
 		VipsPel *p = VIPS_REGION_ADDR( ir, r->left, (int) Y ); 
 		const int sy = Y * VIPS_TRANSFORM_SCALE * 2;
 		const int siy = sy & (VIPS_TRANSFORM_SCALE * 2 - 1);
```

It still produces some pixel shifts (unfortunately), see:
https://github.com/kleisauke/vips-issue-703#shrink-an-image-vertically-1

@jcupitt Do you have any ideas? The whole patch can be found [here](https://github.com/kleisauke/vips-issue-703/blob/master/0001-Partially-fix-for-issue-703.patch). You can safely ignore the `VIPS_DEMAND_STYLE` and `reduce_sum` changes (I tried to align `reduceh` and `reducev`).

**Comment by jcupitt** [comment]
Hello again, I'll have a few hours this weekend, I'll try to have a look then. Well done for digging on this.

**Comment by kleisauke** [comment]
Let me know if you need any assistance (I could perhaps perform a `git bisect`). fwiw, I also checked Pillow's source code, and it looks like they're taking a similar approach (with `+ 0.5` offset):
https://github.com/python-pillow/Pillow/blob/67f309cec8173aff2e069f4683f2861eec3578f6/src/libImaging/Resample.c#L179

But then within the coefficients calculations, so perhaps we'll need to check `calculate_coefficients_lanczos`:
https://github.com/libvips/libvips/blob/29d9673fd8b5558adc0929a35c50974620c1e7a7/libvips/resample/templates.h#L404
(I tried to change values in there, and I saw an increase in the number of pixel shifts)

Also, the above mentioned patch doesn't work when I change the kernel to `VIPS_KERNEL_LINEAR`, so the other kernels need to be checked as well:
https://github.com/libvips/libvips/blob/c7f98ad0a58f23693cf9f36ba8cd6fc8d85352a1/libvips/resample/resize.c#L398

**Comment by kleisauke** [comment]
I just stopped the `git bisect` command that ran for a few hours (the full log is available [here](https://gist.github.com/kleisauke/0c2a47ed19d5a9c4fa275db157070cfe)). I think this bug is present in all versions of libvips starting from commit fa7d5939f5e8462ca045f42854b7706645e18802 (where `vipsthumbnail` was added).

The test script is available [here](https://github.com/kleisauke/vips-issue-703/blob/master/input/unlabelled/test.sh). It's a simple test where the average value is calculated (rounded to one decimal place) after the EXIF orientation image (without labels) is thumbnailed and converted to greyscale. If all images have the same average value, it will exit with 0, otherwise with 125 (which causes `git bisect` to skip that commit).

**Comment by kleisauke** [comment]
fwiw, this bug still seems to be present on the master branch. I'm going to check the coefficients calculations to see if I can find any issues.

**Comment by kleisauke** [comment]
I'm not sure why I didn't test the other resize kernels earlier. `VIPS_KERNEL_NEAREST` doesn't need any resizing coefficients, so it's a useful way of checking whether the coefficients calculations should also be adapted.

I added this kernel (+ the other resize kernels) with commit https://github.com/kleisauke/vips-issue-703/commit/be734f3d7961e0fc79996d62a93f33a4f4a0bfce and regenerated the output before and after applying [this patch](https://github.com/kleisauke/vips-issue-703/blob/master/0001-Partially-fix-for-issue-703.patch):
<details>
  <summary>VIPS_KERNEL_NEAREST - Portrait - before/after</summary>

  ![Portrait before](https://github.com/kleisauke/vips-issue-703/raw/be734f3d7961e0fc79996d62a93f33a4f4a0bfce/output/nearest_Portrait.webp)
  ![Portrait after](https://github.com/kleisauke/vips-issue-703/raw/be734f3d7961e0fc79996d62a93f33a4f4a0bfce/output-patch/nearest_Portrait.webp)
</details>
<details>
  <summary>VIPS_KERNEL_NEAREST - Landscape - before/after</summary>

  ![Landscape before](https://github.com/kleisauke/vips-issue-703/raw/be734f3d7961e0fc79996d62a93f33a4f4a0bfce/output/nearest_Landscape.webp)
  ![Landscape after](https://github.com/kleisauke/vips-issue-703/raw/be734f3d7961e0fc79996d62a93f33a4f4a0bfce/output-patch/nearest_Landscape.webp)
</details>

At first sight, it looks like they're all shifting. However, if you look more closely (try to zoom-in 500%) the landscape before image is the only image that is affected by pixel shifts. All other images are not affected. This implicitly means that the (partial) patch is working, but in order to solve this issue completely, a change in coefficients calculations is also necessary.

To be continued (I've some spare time tomorrow).

**Comment by kleisauke** [comment]
I managed to (partially) fix the cubic, linear and Mitchell kernel with commit https://github.com/kleisauke/vips-issue-703/commit/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784.

The idea was to first calculate the kernel radius (rounded to the nearest integer). This integer is then used:
- to calculate the number of points within the kernel (usually `2 * kernel_radius + 1`);
- within the coefficients calculations (usually `(i - kernel_radius - x) / shrink`);
- for adding new pixels to the image before reducing (so we can interpolate at the edges)

The whole patch can be found [here](https://github.com/kleisauke/vips-issue-703/blob/master/0001-Partially-fix-for-issue-703.patch). Here are the before/after results for these kernels:
<details>
  <summary>VIPS_KERNEL_LINEAR - before/after</summary>

**Landscape before:**
  ![Landscape before](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output/linear_Landscape.webp)
**Landscape after:**
  ![Landscape after](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output-patch/linear_Landscape.webp)
**Portrait before:**
  ![Portrait before](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output/linear_Portrait.webp)
**Portrait after:**
  ![Portrait after](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output-patch/linear_Portrait.webp)
</details>
<details>
  <summary>VIPS_KERNEL_CUBIC - before/after</summary>

**Landscape before:**
  ![Landscape before](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output/cubic_Landscape.webp)
**Landscape after:**
  ![Landscape after](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output-patch/cubic_Landscape.webp)
**Portrait before:**
  ![Portrait before](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output/cubic_Portrait.webp)
**Portrait after:**
  ![Portrait after](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output-patch/cubic_Portrait.webp)
</details>
<details>
  <summary>VIPS_KERNEL_MITCHELL - before/after</summary>

**Landscape before:**
  ![Landscape before](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output/mitchell_Landscape.webp)
**Landscape after:**
  ![Landscape after](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output-patch/mitchell_Landscape.webp)
**Portrait before:**
  ![Portrait before](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output/mitchell_Portrait.webp)
**Portrait after:**
  ![Portrait after](https://github.com/kleisauke/vips-issue-703/raw/19b628e4dab7f86a5cc6fe9a191f7b4c9550a784/output-patch/mitchell_Portrait.webp)
</details>

All portrait images seems to be properly fixed. The landscape images still produce a small pixel shift (unfortunately). This can be clearly spotted if an image is only vertically reduced:
https://raw.githubusercontent.com/kleisauke/vips-issue-703/master/output-patch/Landscape-vertical.webp

TL;DR: the only thing that needs to be resolved is the small pixel shift within `reducev`.

**Comment by jcupitt** [comment]
This should be fixed in 8.10. I'll close.
